edward/inferences/variational_inference.py

```diff
@@ -53,25 +53,70 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
     """
     super(VariationalInference, self).initialize(*args, **kwargs)
 
+    latent_var_list = set()
+    data_var_list = set()
     if var_list is None:
       # Traverse random variable graphs to get default list of variables.
-      var_list = set()
       trainables = tf.trainable_variables()
       for z, qz in six.iteritems(self.latent_vars):
         if isinstance(z, RandomVariable):
-          var_list.update(get_variables(z, collection=trainables))
+          latent_var_list.update(get_variables(z, collection=trainables))
 
-        var_list.update(get_variables(qz, collection=trainables))
+        latent_var_list.update(get_variables(qz, collection=trainables))
 
       for x, qx in six.iteritems(self.data):
         if isinstance(x, RandomVariable) and \
                 not isinstance(qx, RandomVariable):
-          var_list.update(get_variables(x, collection=trainables))
+          data_var_list.update(get_variables(x, collection=trainables))
 
-      var_list = list(var_list)
+      var_list = list(data_var_list | latent_var_list)
 
     self.loss, grads_and_vars = self.build_loss_and_gradients(var_list)
 
+    if self.logging:
+      summary_key = 'summaries_' + str(id(self))
+      tf.summary.scalar("loss", self.loss, collections=[summary_key])
+      with tf.name_scope('variational'):
+        for grad, var in grads_and_vars:
+          if var in latent_var_list:
+            tf.summary.histogram("parameter_" +
+                                 var.name.replace(':', '_'),
+                                 var, collections=[summary_key])
+            tf.summary.histogram("gradient_" +
+                                 grad.name.replace(':', '_'),
+                                 grad, collections=[summary_key])
+            tf.summary.scalar("gradient_norm_" +
+                              grad.name.replace(':', '_'),
+                              tf.norm(grad), collections=[summary_key])
+      # replace : with _ because tf does not allow : in var names in summaries
```

Comment:
1. naming of gradients messy: see comment above

2. tracking gradients is super useful, but it might be even more useful to look at `gradients*learning_rate` or `(parameters - gradients*learning_rate)/parameters`. 
Especially with adaptive learning rates when the user doesn't know the current learning rate, it is impossible to know if the gradient is in the right ballpark from looking at the histograms.

3. when training blows up and some parameters become `NaN` or `inf` (e.g. because the learning rate is too large) I get an error message that is hard to decipher.
```
InvalidArgumentError (see above for traceback): Nan in summary histogram for: model_1/parameter_model/embeddings/alpha_0
         [[Node: model_1/parameter_model/embeddings/alpha_0 = HistogramSummary[T=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"](model_1/parameter_model/embeddings/alpha_0/tag, model/embeddings/alpha/read)]]
```
Maybe there's a nicer way to catch this? I guess that's more an idiosyncrasy of tensorflow than of edward. But still, catching it in a more readable way would be nice